qiskit_experiments/framework/experiment_data.py

```diff
@@ -767,6 +770,66 @@ def add_data(
     ) -> None:
         """Add experiment data.
 
+        Args:
+            data: Experiment data to add. Several types are accepted for convenience:
+
+                * Result: Add data from this ``Result`` object.
+                * List[Result]: Add data from the ``Result`` objects.
+                * Dict: Add this data.
+                * List[Dict]: Add this list of data.
+
+        Raises:
+            TypeError: If the input data type is invalid.
+        """
+
+        if any(not future.done() for future in self._analysis_futures.values()):
+            LOG.warning(
+                "Not all analysis has finished running. Adding new data may "
+                "create unexpected analysis results."
+            )
+        if not isinstance(data, list):
+            data = [data]
+
+        # Directly add non-job data
+        with self._result_data.lock:
+            tmp_exp_data = ExperimentData()
+            composite_flag = False
+            experiment_seperator = {}
+            for datum in data:
+                if isinstance(datum, dict):
+                    if "composite_metadata" in datum["metadata"]:
+                        composite_flag = True
+                        marginalized_data = self._marginalized_component_data([datum])
+                        for inner_datum in marginalized_data:
+                            #print(inner_datum)
+                            if "experiment_type" in inner_datum[0]["metadata"]:
+                                if inner_datum[0]["metadata"]["experiment_type"] in experiment_seperator:
+                                    experiment_seperator[inner_datum[0]["metadata"]["experiment_type"]].add_data(inner_datum[0])
+                                else:
+                                    experiment_seperator[inner_datum[0]["metadata"]["experiment_type"]] = ExperimentData()
+                                    experiment_seperator[inner_datum[0]["metadata"]["experiment_type"]].add_data(inner_datum[0])
+
+                        self._result_data.append(datum)
+
+                elif isinstance(datum, Result):
+                    if datum["metadata"]:
+                        self._set_child_data(datum["metadata"]._metadata())
+                    else:
+                        self._add_result_data(datum)
+                else:
+                    raise TypeError(f"Invalid data type {type(datum)}.")
+            if composite_flag:
+                tmp_exp_data._set_child_data(list(experiment_seperator.values()))
+                self._set_child_data([tmp_exp_data])
+        
+        return tmp_exp_data
```

Comment:
This is creating new object. Instead of preparing and returning new `tmp_exp_data`, you need to update `self._result_data`. Note that `ExperimentData` receives job object from an experiment, and it extracts result data by itself. This means the running experiment data instance should mutate itself through the `.add_data()` method.

While investigating the codebase, I found another place we need to update. Let me explain step by step. When the experiment class creates a Qiskit Job, the `ExpeirmentData` receives the result data with the following steps

- Call `.add_jobs()` method, and it creates a future object with a callback `._add_job_future()` to wait for result data at [this line](https://github.com/Qiskit-Extensions/qiskit-experiments/blob/4897eb59af4eae1e5ad1c6b130265c8d6ed4c236/qiskit_experiments/framework/experiment_data.py#L857).
- This callback further submits another callback `._add_job_data()` to the executor, that extract data from result.
- In the `._add_job_data()` callback, it calls another callback `._add_result_data()` at [this line](https://github.com/Qiskit-Extensions/qiskit-experiments/blob/4897eb59af4eae1e5ad1c6b130265c8d6ed4c236/qiskit_experiments/framework/experiment_data.py#L917).
- Finally, in the `._add_result_data()` callback, it directly updates `self._result_data` at [this line](https://github.com/Qiskit-Extensions/qiskit-experiments/blob/4897eb59af4eae1e5ad1c6b130265c8d6ed4c236/qiskit_experiments/framework/experiment_data.py#L1068).

The implementation is quite complicated right now... But please note that the final step should call this `.add_data()` method with the list of dictionary, instead of appending `data` in each loop. This allows `ExperimentData` to bootstrap result data for component analysis.